Add Sort feature

kkangs0226 opened the pull request:
List of projects displayed cannot be sorted in a specified order.

Adding sort feature will make application more user friendly as it
can display projects in orders required by users.

Let's
* add SortCommand and SortCommandParser
* add other relevant Comparators for sorting
* make necessary changes to existing files
* add tests for SortCommand, SortCommandParser and Comparators

Closes #47, closes #48

kkangs0226 committed:
Add Sort feature

List of projects displayed cannot be sorted in a specified order.

Adding sort feature will make application more user friendly as it
can display projects in orders required by users.

Let's
* add SortCommand and SortCommandParser
* add other relevant Comparators for sorting
* make necessary changes to existing files
* add tests for SortCommand, SortCommandParser and Comparators

kkangs0226 committed:
Merge branch 'master' into feature-issue-47

kkangs0226 committed:
Merge branch 'master' of https://github.com/AY2021S1-CS2103T-T10-1/tp into feature-issue-47

kkangs0226 committed:
Update User Guide

User Guide does not have a section for sort command.

Updating user guide to include sort instructions would improve
user guide.

Let's update user guide to include sort instructions.

kkangs0226 committed:
Merge branch 'feature-issue-47' of https://github.com/kkangs0226/tp into feature-issue-47

kkangs0226 committed:
Fix accidental change in User Guide

Accidentally removed a character from User Guide.

Let's fix the change.

khoodehui commented on src/main/java/seedu/momentum/logic/commands/SortCommand.java:
```diff
@@ -0,0 +1,92 @@
+
+    /**
+     * Creates a SortCommand to sort the list of projects.
+     * @param sortType Type of sort applied to projects.
```
Newline before param

kkangs0226 replied:
Somehow the checkstyle doesn't pass it if i leave a newline, I think we can disregard these for now?

khoodehui commented on src/main/java/seedu/momentum/logic/parser/SortCommandParser.java:
```diff
@@ -0,0 +1,78 @@
+
+    /**
+     * Parses the given {@code String} of arguments in the context of the SortCommand
+     * and returns a SortCommand object for execution.
```
Newline after description

kkangs0226 replied:
Same as my previous comment

pr4aveen commented on src/main/java/seedu/momentum/logic/commands/SortCommand.java:
```diff
@@ -0,0 +1,92 @@
+            + "Ascending: asc; Descending: dsc. \n"
+            + "Example: " + COMMAND_WORD + " " + SORT_TYPE + "alpha " + SORT_ORDER + "asc";
+
+    public static final String MESSAGE_SORT_SUCCESS = "List has been sorted in %1$s%2$s order";
```
Should this go under `seedu.momentum.commons.core.Messages`?

kkangs0226 replied:
I think we can leave it here since other classes are doing the same. We can move them to Messages in one go in the next iteration!

pr4aveen commented on src/main/java/seedu/momentum/model/ModelManager.java:
```diff
@@ -35,6 +39,9 @@ public ModelManager(ReadOnlyProjectBook projectBook, ReadOnlyUserPrefs userPrefs
         this.projectBook = new ProjectBook(projectBook);
         this.userPrefs = new UserPrefs(userPrefs);
         filteredProjects = new FilteredList<>(this.projectBook.getProjectList());
+        this.currentPredicate = PREDICATE_SHOW_ALL_PROJECTS;
+        this.currentSortType = SortType.ALPHA;
+        this.currentSortIsAscending = true;
```
Should these 3 variables just be declared directly since their values aren't passed into the constructor?

kkangs0226 replied:
Fixed!

pr4aveen commented on src/main/java/seedu/momentum/model/project/comparators/CreatedDateCompare.java:
```diff
@@ -0,0 +1,32 @@
+
+        NameCompare nameCompare = new NameCompare();
+
+        if (p1Date.get().isBefore(p2Date.get())) {
+            return -1;
+        } else if (p1Date.get().isAfter(p2Date.get())) {
+            return 1;
```
Should we implement a `compareTo` method in `Date.java`?

khoodehui replied:
I think we should.

kkangs0226 replied:
Done!

pr4aveen commented on src/main/java/seedu/momentum/model/project/comparators/DeadlineCompare.java:
```diff
@@ -0,0 +1,60 @@
+import seedu.momentum.commons.core.Time;
+import seedu.momentum.model.project.Deadline;
+
+/**
+ * Compar
+ */
```
Incomplete header here

kkangs0226 replied:
Fixed!

pr4aveen commented on src/main/java/seedu/momentum/model/project/comparators/DeadlineCompare.java:
```diff
@@ -0,0 +1,60 @@
+     */
+    public int compare(HashMap<String, Object> p1, HashMap<String, Object> p2) {
+        Deadline p1Deadline = (Deadline) p1.get("deadline");
+        Date p1DeadlineDate = p1Deadline.getDate();
+        Deadline p2Deadline = (Deadline) p2.get("deadline");;
+        Date p2DeadlineDate = p2Deadline.getDate();
+
+        String p1Name = (String) p1.get("name");
+        String p2Name = (String) p2.get("name");
+
+        if (p1DeadlineDate.get().isBefore(p2DeadlineDate.get())) {
+            return -1;
+        } else if (p1DeadlineDate.get().isAfter(p2DeadlineDate.get())) {
+            return 1;
+        } else {
+            if (!p1Deadline.hasTime() && p2Deadline.hasTime()) {
+                return -1;
+            } else if (p1Deadline.hasTime() && !p2Deadline.hasTime()) {
+                return 1;
+            } else if (p1Deadline.hasTime() && p2Deadline.hasTime()) {
+                return bothHasTimeCompare(p1Deadline, p1Name, p2Deadline, p2Name);
+            } else {
+                return p1Name.compareTo(p2Name);
+            }
+        }
+    }
+
+    private int bothHasTimeCompare(Deadline p1Deadline, String p1Name, Deadline p2Deadline, String p2Name) {
+        Time p1DeadlineTime = p1Deadline.getTime();
+        Time p2DeadlineTime = p2Deadline.getTime();
+
+        if (p1DeadlineTime.get().isBefore(p2DeadlineTime.get())) {
+            return -1;
+        } else if (p1DeadlineTime.get().isAfter(p2DeadlineTime.get())) {
+            return 1;
+        } else {
+            return p1Name.compareTo(p2Name);
+        }
+    }
+}
```
I think we should have `compareTo` methods in the `Date` and `Time` classes and use them here instead

kkangs0226 replied:
Done as well!

pr4aveen commented on src/main/java/seedu/momentum/model/project/Project.java:
```diff
@@ -90,6 +91,23 @@ public Deadline getDeadline() {
         return deadline;
     }
 
+    /**
+     * Gets Deadline and name for {@code DeadLineCompare} comparator.
+     * Returns null if deadline is empty as {@code Comparator.nullsLast} method is used.
+     */
+    public HashMap<String, Object> getNullOrDeadline() {
+        if (deadline.isEmpty()) {
+            return null;
+        }
+        String name = getName().fullName;
+        Deadline deadline = getDeadline();
+        HashMap<String, Object> nameDeadlineMap = new HashMap<>();
+        nameDeadlineMap.put("name", name);
+        nameDeadlineMap.put("deadline", deadline);
+
+        return nameDeadlineMap;
```
I'm not sure if there's a cleaner way to do this. What do the rest think?

kkangs0226 replied:
The reason why I need to use HashMap is because not all projects have Deadlines, and I'm using Comparator.nullsLast to place all projects without deadlines at the end of the list after projects with deadlines, so I need to compare Deadlines (and not Projects). But after extracting the deadlines (or nulls), I will still need to compare the names alphabetically in case of same deadline or null deadline, which is why I need to return the name as well. Anyone can lmk if there is a cleaner way to do this!

pr4aveen commented on src/main/java/seedu/momentum/model/project/comparators/DeadlineCompare.java:
```diff
@@ -0,0 +1,60 @@
+     * @return integer values for comparison.
+     */
+    public int compare(HashMap<String, Object> p1, HashMap<String, Object> p2) {
+        Deadline p1Deadline = (Deadline) p1.get("deadline");
+        Date p1DeadlineDate = p1Deadline.getDate();
+        Deadline p2Deadline = (Deadline) p2.get("deadline");;
+        Date p2DeadlineDate = p2Deadline.getDate();
+
+        String p1Name = (String) p1.get("name");
+        String p2Name = (String) p2.get("name");
```
This part is related to the comment on `Project.java`

pr4aveen commented on src/test/java/seedu/momentum/logic/commands/SortCommandTest.java:
```diff
@@ -0,0 +1,149 @@
+    public void execute_deadlineDescending_sortedInDeadlineDescendingOrder() {
+        expectedModel.orderFilteredProjectList(SortType.DEADLINE, false);
+        String expectedMessage = String.format(MESSAGE_SORT_SUCCESS, "deadline, ", "descending");
+        assertCommandSuccess(DEADLINE_DESCENDING, model, expectedMessage, expectedModel);
+        assertEquals(model.getFilteredProjectList(), expectedModel.getFilteredProjectList());
+    }
+
+    @Test
+    public void execute_createdDateAscending_sortedInCreatedDateAscendingOrder() {
+        expectedModel.orderFilteredProjectList(SortType.CREATED, true);
+        String expectedMessage = String.format(MESSAGE_SORT_SUCCESS, "created date, ", "ascending");
+        assertCommandSuccess(CREATED_DATE_ASCENDING, model, expectedMessage, expectedModel);
+        assertEquals(model.getFilteredProjectList(), expectedModel.getFilteredProjectList());
+    }
+
+    @Test
+    public void execute_createdDateDescending_sortedInCreatedDateDescendingOrder() {
+        expectedModel.orderFilteredProjectList(SortType.CREATED, false);
+        String expectedMessage = String.format(MESSAGE_SORT_SUCCESS, "created date, ", "descending");
+        assertCommandSuccess(CREATED_DATE_DESCENDING, model, expectedMessage, expectedModel);
+        assertEquals(model.getFilteredProjectList(), expectedModel.getFilteredProjectList());
+    }
+
+    @Test
+    public void execute_nullSortTypeAscendingNonDefault_sortedInCurrentSortTypeAscendingOrder() {
+        model.orderFilteredProjectList(SortType.DEADLINE, false);
+        expectedModel.orderFilteredProjectList(SortType.DEADLINE, true);
+        String expectedMessage = String.format(MESSAGE_SORT_SUCCESS, "", "ascending");
+        assertCommandSuccess(NULL_SORT_TYPE_ASCENDING_NON_DEFAULT, model, expectedMessage, expectedModel);
+        assertEquals(model.getFilteredProjectList(), expectedModel.getFilteredProjectList());
+    }
+
+    @Test
+    public void execute_nullSortTypeDescendingNonDefault_sortedInCurrentSortTypeDscendingOrder() {
+        model.orderFilteredProjectList(SortType.CREATED, true);
+        expectedModel.orderFilteredProjectList(SortType.CREATED, false);
+        String expectedMessage = String.format(MESSAGE_SORT_SUCCESS, "", "descending");
+        assertCommandSuccess(NULL_SORT_TYPE_DESCENDING_NON_DEFAULT, model, expectedMessage, expectedModel);
+        assertEquals(model.getFilteredProjectList(), expectedModel.getFilteredProjectList());
+    }
```
I think you can extract the common string variables that you have reused in this section and make them private static variables. For example, `ascending`, `descending`, `alphabetical`, `deadline`, `created date`

kkangs0226 replied:
Fixed!

pr4aveen commented on src/main/java/seedu/momentum/logic/parser/SortCommandParser.java:
```diff
@@ -0,0 +1,78 @@
+        sortType = sortType.trim();
+
+        switch (sortType) {
+        case "alpha":
+            return SortType.ALPHA;
+        case "deadline":
+            return SortType.DEADLINE;
+        case "created":
+            return SortType.CREATED;
```
You might want to make these public static variables instead. It's used quite a fair bit in tests and it'll be better if they all refer to one variable

kkangs0226 replied:
Fixed!

pr4aveen commented on src/test/java/seedu/momentum/logic/parser/ProjectBookParserTest.java:
```diff
@@ -77,6 +81,15 @@ public void parseCommand_find() throws Exception {
+        assertTrue(parser.parseCommand(SortCommand.COMMAND_WORD) instanceof SortCommand);
+        assertTrue(parser.parseCommand(SortCommand.COMMAND_WORD + " 3") instanceof SortCommand);
+        SortCommand command = (SortCommand) parser.parseCommand(SortCommand.COMMAND_WORD + " "
+                + SORT_ORDER + SortCommand.ASCENDING_ORDER + " " + SORT_TYPE + "alpha");
```
Might want to replace the `alpha` (and other keywords in other classes) with the public static variables I recommended in `SortCommandParser.java`

kkangs0226 replied:
Fixed!

pr4aveen commented on src/main/java/seedu/momentum/logic/parser/SortCommandParser.java:
```diff
@@ -0,0 +1,78 @@
+     * and returns a SortCommand object for execution.
+     * @throws ParseException if the user does not conform to the expected format.
+     */
+    public SortCommand parse(String args) throws ParseException {
```
Just checking, what's the expected behaviour when there is a preamble?

For example, `sort hello type/alpha order/dsc`

kkangs0226 replied:
It should disregard the preamble. 
I have added a test for this!

kkangs0226 replied:
I have changed the SortCommandParser to fail when there is a non-empty preamble for consistency as AddCommandParser and FindCommandParser does the same. I have added a test accordingly!

pr4aveen commented on src/test/java/seedu/momentum/logic/commands/AddCommandIntegrationTest.java:
```diff
@@ -36,6 +37,23 @@ public void execute_newProject_success() {
+        Project dana = new ProjectBuilder().withName("Dana").build();
+        AddCommand addDanaCommand = new AddCommand(dana);
+
+        // alphabetical order -> Hello gets placed in between Carl and Daniel
```
Is `Hello` referring to `Dana`?

kkangs0226 replied:
Yes I'll go fix it!

pr4aveen reviewed (approved):
Have a few comments that might need to be addressed before merging. Should be ok to merge once those changes have been made :)

khoodehui commented on src/main/java/seedu/momentum/logic/parser/SortCommandParser.java:
```diff
@@ -0,0 +1,78 @@
+        ArgumentMultimap argMultimap =
+                ArgumentTokenizer.tokenize(args, SORT_TYPE, SORT_ORDER);
+
+        boolean isAscending = parseSortOrder(argMultimap);
+        SortType sortType = parseSortType(argMultimap);
+        boolean isDefault = false;
+
+        if (argMultimap.getValue(SORT_TYPE).isEmpty() && argMultimap.getValue(SORT_ORDER).isEmpty()) {
+            isDefault = true;
+            sortType = SortType.NULL;
+        }
+
+        return new SortCommand(sortType, isAscending, isDefault);
+    }
+
+    private boolean parseSortOrder(ArgumentMultimap argMultimap) throws ParseException {
+        if (argMultimap.getValue(SORT_ORDER).isEmpty()) {
+            return true;
+        }
+
+        String sortOrder = argMultimap.getValue(SORT_ORDER).get();
+        sortOrder = sortOrder.trim();
+
+        switch (sortOrder) {
+        case ASCENDING_ORDER:
+            return true;
+        case DESCENDING_ORDER:
+            return false;
+        default:
+            throw new ParseException(String.format(
+                    MESSAGE_INVALID_COMMAND_FORMAT, SortCommand.MESSAGE_INVALID_SORT_TYPE_OR_ORDER));
+        }
+    }
```
Code might be easier to understand if `parseSortOrder()` only returns the argument which the user entered in the command. Then under the variable declaration for `isAscending()`, do an `equal()` check to see whether the argument the user provided is equal to `ASCENDING_ORDER`.

This is so that the nature of the `parseSortOrder()` method is also consistent with that of the `parseSortType()` method.

kkangs0226 replied:
Fixed!

khoodehui commented on src/test/java/seedu/momentum/logic/parser/SortCommandParserTest.java:
```diff
@@ -0,0 +1,84 @@
+    public void parse_missingSortOrder_returnsTypedAscendingSortCommand() {
+
+        // Alphabetical type
+        SortCommand expectedCommand = new SortCommand(SortType.ALPHA, true, false);
+        assertParseSuccess(parser, VALID_ALPHA_SORT_TYPE, expectedCommand);
+
+        // Deadline type
+        expectedCommand = new SortCommand(SortType.DEADLINE, true, false);
+        assertParseSuccess(parser, VALID_DEADLINE_SORT_TYPE, expectedCommand);
+
+        // Created date type
+        expectedCommand = new SortCommand(SortType.CREATED, true, false);
+        assertParseSuccess(parser, VALID_CREATED_DATE_SORT_TYPE, expectedCommand);
```
Noticed that you have declared various form of `SortCommands` as private static final variables in `SortCommandTest.java`, and they are repeated here. Will it be better to just abstract all of the various `SortCommands` into a utility class, then have the test classes get the different commands from there.

kkangs0226 replied:
Abstracted into a SortCommandUtil class!

khoodehui reviewed (approved):
Just some things to take note, not too much of an issue though. Otherwise LGTM!

claracheong4 commented on src/main/java/seedu/momentum/model/Model.java:
```diff
@@ -85,4 +86,10 @@
     void updateFilteredProjectList(Predicate<Project> predicate);
+
+    /**
+     * Orders the list of projects in a way given by the {@code sortType}.
```
New line needed before @throws

kkangs0226 replied:
Fixed!

claracheong4 commented on src/main/java/seedu/momentum/model/project/UniqueProjectList.java:
```diff
@@ -97,6 +104,69 @@ public void setProjects(List<Project> projects) {
+
+        requireNonNull(sortType);
+
+        Comparator<Project> nameCompare = new NameCompare();
```
Make name implement a comparable instead?

kkangs0226 replied:
Done!

claracheong4 commented on src/main/java/seedu/momentum/model/project/UniqueProjectList.java:
```diff
@@ -97,6 +104,69 @@ public void setProjects(List<Project> projects) {
+
+        switch (sortType) {
+        case ALPHA:
+            nameCompare = isAscending ? nameCompare : nameCompare.reversed();
+            this.sortType = SortType.ALPHA;
+            internalList.sort(nameCompare);
```
The method is quite long, maybe extract this into another method?

kkangs0226 replied:
Fixed!

claracheong4 commented on src/main/java/seedu/momentum/model/project/UniqueProjectList.java:
```diff
@@ -97,6 +104,69 @@ public void setProjects(List<Project> projects) {
+            internalList.sort(nameCompare);
+            break;
+        case DEADLINE:
+            Comparator<HashMap<String, Object>> deadlineCompare = new DeadlineCompare();
+            deadlineCompare = isAscending ? deadlineCompare : deadlineCompare.reversed();
+            this.sortType = SortType.DEADLINE;
+            internalList.sort(Comparator.comparing(Project::getNullOrDeadline, Comparator.nullsLast(deadlineCompare))
+                    .thenComparing(nameCompare));
```
similarly for this

kkangs0226 replied:
Fixed!

claracheong4 commented on src/main/java/seedu/momentum/model/project/UniqueProjectList.java:
```diff
@@ -97,6 +104,69 @@ public void setProjects(List<Project> projects) {
+                    .thenComparing(nameCompare));
+            break;
+        case CREATED:
+            Comparator<Project> createdDateCompare = new CreatedDateCompare();
+            createdDateCompare = isAscending ? createdDateCompare : createdDateCompare.reversed();
+            this.sortType = SortType.CREATED;
+            internalList.sort(createdDateCompare);
```
and this?

kkangs0226 replied:
Fixed!

claracheong4 reviewed (commented):
I think some minor changes has to be made.

kkangs0226 committed:
Implement compareTo methods in model classes and update code to
resolve comments

Name, Deadline, Date, DateTime, Time classes can inherit from
Comparable interface to make them comparable.
Some long methods can be abstracted further and more static
variables can be declared to make code more readable.

Let's
* implement compareTo methods in these classes
* update code to make it more readable

kkangs0226 committed:
Merge branch 'master' of https://github.com/AY2021S1-CS2103T-T10-1/tp into feature-issue-47

kkangs0226 committed:
Update import order for ProjectBookParserTest

Wrong import order causes checkstyle error.

Let's update import order.

kkangs0226 committed:
Update SortCommandParserTest

Test for parsing empty arguments fails on Windows.

Let's update SortCommandParserTest.

kkangs0226 committed:
Second update to SortCommandParserTest

Test for parsing empty arguments fails on Windows.

Let's update SortCommandParserTest.

kkangs0226 committed:
Add tests for compareTo methods for model classes

compareTo methods have not been tested.

Add tests to increase test code coverage.

Let's add tests for compareTo methods.
